### PR TITLE
Various CR fixes and EoU enhancements for validate

### PR DESCRIPTION
**Pull request by AShivangi**
- CR-1099874 xbutil2 validate fails on xilinx_u250_gen3x16_xdma_shell_3_1 after the slave bridge is enabled (Fix: host mem check)
- CR-1099470 xbutil valdiate fail after call "xbutil hostmem --enable --size 512M" for xilinx_u200_gen3x16_xdma_1_202110_1 (Dup of above)
- CR-1099860 xbutil2 validate on xilinx_u250_gen3x16_xdma_shell_3_1 shows unclear message about vcu test, which is not supported (Fix: add a message at the end asking the user to run --verbose for more details. A much bigger cleanup will come after 21.1 release)
- CR-1099056 xbutil2 validate bandwidth kernel times out on U50 gen3x16 shells (Fix: increase timeout to 2 minutes)
- CR-1097540 new xrt tool sc flash behavior needs to improve (Workaround: ask the user to do a warm reboot)
- Enforce user to specify output file when "--format" is specified

Sample outputs:
DMA test when host mem is enabled & 'Please run the command '--verbose' option for more details' message
```
% Debug/opt/xilinx/xrt/bin/xbutil validate -d 65:00 -r 'DMA'
Starting validation for 1 devices

Validate Device           : [0000:65:00.1]
    Platform              : xilinx_u250_gen3x16_xdma_shell_3_1
    SC Version            : 
    Platform ID           : 0x0
-------------------------------------------------------------------------------
Test 1 [0000:65:00.1]     : DMA 
    Details               : Host -> PCIe -> FPGA write bandwidth = 9463.452728 MB/s
                            Host <- PCIe <- FPGA read bandwidth = 11775.021201 MB/s
    Test Status           : [PASSED]
-------------------------------------------------------------------------------
Validation completed. Please run the command '--verbose' option for more details
```
Specify output file:
```
$ xbutil examine -d 17:00 -f json
ERROR: Please specify an output file to redirect the json to
```

### CODE REVIEW

File: src/runtime_src/core/tools/xbmgmt2/SubCmdExamine.cpp [hunk 1]
```diff
@@ -135,6 +135,10 @@ SubCmdExamine::execute(const SubCmdOptions& _options) const
     if (schemaVersion == Report::SchemaVersion::unknown) 
       throw xrt_core::error((boost::format("Unknown output format: '%s'") % sFormat).str());
 
+    // when json is specified, make sure an accompanying output file is also specified
+    if (!vm["format"].defaulted() && sOutput.empty())
```
Comment (rozumx) [inline src/runtime_src/core/tools/xbmgmt2/SubCmdExamine.cpp:139]: 👍

File: src/runtime_src/core/tools/xbmgmt2/SubCmdProgram.cpp [hunk 2]
```diff
@@ -465,6 +468,10 @@ auto_flash(xrt_core::device_collection& deviceCollection)
     std::cout << "****************************************************\n";
     std::cout << "Cold reboot machine to load the new image on device(s).\n";
     std::cout << "****************************************************\n";
+  } else if (need_warm_reboot) {
+    std::cout << "****************************************************\n";
+    std::cout << "Warm reboot machine to load the new SC image on device(s).\n";
```
Comment (rozumx) [inline src/runtime_src/core/tools/xbmgmt2/SubCmdProgram.cpp:473]: Wording should be improved.  Something like:
"Warm reboot of the device is required to recognize new SC image."

**Review by rozumx — CHANGES_REQUESTED** [review]

File: src/runtime_src/core/tools/xbmgmt2/SubCmdProgram.cpp [hunk 3]
```diff
@@ -434,16 +436,17 @@ auto_flash(xrt_core::device_collection& deviceCollection)
 
     // Perform DSA and BMC updating
     for (auto& p : boardsToUpdate) {
-      bool reboot;
+      bool reboot, warm_reboot;
```
Comment (stsoe) [inline src/runtime_src/core/tools/xbmgmt2/SubCmdProgram.cpp:439]: warning: variable `reboot` is not initialized [cppcoreguidelines-init-variables]

```cpp
bool reboot, warm_reboot;
     ^
            = 0
```

Comment (stsoe) [inline src/runtime_src/core/tools/xbmgmt2/SubCmdProgram.cpp:439]: warning: variable `warm_reboot` is not initialized [cppcoreguidelines-init-variables]

```cpp
bool reboot, warm_reboot;
             ^
                         = 0
```

**Review by stsoe — COMMENTED** [review]
clang-tidy made some suggestions

File: src/runtime_src/core/tools/xbmgmt2/SubCmdProgram.cpp [hunk 4]
```diff
@@ -326,7 +326,7 @@ isSameShellOrSC(const DSAInfo& candidate, const DSAInfo& current, bool& same_dsa
  * Helper method for auto_flash
  */
 static int 
-updateShellAndSC(unsigned int  boardIdx, DSAInfo& candidate, bool& reboot)
+updateShellAndSC(unsigned int  boardIdx, DSAInfo& candidate, bool& reboot, bool& warm_reboot)
 {
   reboot = false;
```
Comment (rozumx) [inline src/runtime_src/core/tools/xbmgmt2/SubCmdProgram.cpp:331]: Being that warm_reboot is a return variable, shouldn't it be initialized first to a known value?  
Something like:
`warm_reboot = false;`

File: src/runtime_src/core/tools/xbmgmt2/SubCmdProgram.cpp (hunk 3 above)
Comment (rozumx) [reply src/runtime_src/core/tools/xbmgmt2/SubCmdProgram.cpp:439]: C++ Core Guidelines: https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines.html#Res-always

Comment (rozumx) [inline src/runtime_src/core/tools/xbmgmt2/SubCmdProgram.cpp:439]: While:
`bool reboot, warm_reboot = false;  `
is valid, for readability and better refactoring support,  it is better to declare each variable uniquely.  For example:
```
bool reboot = false;
bool warm_reboot = false;
```

File: src/runtime_src/core/tools/xbutil2/SubCmdExamine.cpp [hunk 5]
```diff
@@ -114,7 +114,7 @@ SubCmdExamine::execute(const SubCmdOptions& _options) const
   commonOptions.add_options()
     ("device,d", boost::program_options::value<decltype(devices)>(&devices)->multitoken(), "The Bus:Device.Function (e.g., 0000:d8:00.0) device of interest.  A value of 'all' (default) indicates that every found device should be examined.")
     ("report,r", boost::program_options::value<decltype(reportNames)>(&reportNames)->multitoken(), (std::string("The type of report to be produced. Reports currently available are:\n") + reportOptionValues).c_str() )
-    ("format,f", boost::program_options::value<decltype(sFormat)>(&sFormat), (std::string("Report output format. Valid values are:\n") + formatOptionValues).c_str() )
+    ("format,f", boost::program_options::value<decltype(sFormat)>(&sFormat)->default_value("json"), (std::string("Report output format. Valid values are:\n") + formatOptionValues).c_str() )
```
Comment (rozumx) [inline src/runtime_src/core/tools/xbutil2/SubCmdExamine.cpp:117]: Hmm...this doesn't look right.  Why is there a default value here?

See comment below for further detail.

File: src/runtime_src/core/tools/xbutil2/SubCmdExamine.cpp [hunk 6]
```diff
@@ -172,6 +172,12 @@ SubCmdExamine::execute(const SubCmdOptions& _options) const
     throw xrt_core::error(std::errc::operation_canceled);
   }
 
+  // when json is specified, make sure an accompanying output file is also specified
+  if (!vm["format"].defaulted() && sOutput.empty()) {
```
Comment (rozumx) [inline src/runtime_src/core/tools/xbutil2/SubCmdExamine.cpp:176]: Should this code be as follows:
```suggestion
  if (!sFormat.empty() && sOutput.empty()) {
```

File: src/runtime_src/core/tools/xbutil2/SubCmdValidate.cpp [hunk 7]
```diff
@@ -842,8 +842,15 @@ dmaTest(const std::shared_ptr<xrt_core::device>& _dev, boost::property_tree::ptr
       break;
   }
 
+  auto is_host_mem = [](std::string tag) {
+    return tag.compare(0,4,"HOST") == 0;
```
Comment (rozumx) [inline src/runtime_src/core/tools/xbutil2/SubCmdValidate.cpp:846]: This line pretty much doesn't do anything for this function doe not return a value.

**Review by rozumx — CHANGES_REQUESTED** [review]
Changes requested.

**Review by rozumx — CHANGES_REQUESTED** [review]

File: src/runtime_src/core/tools/xbutil2/SubCmdValidate.cpp (hunk 7 above)
Comment (rozumx) [reply src/runtime_src/core/tools/xbutil2/SubCmdValidate.cpp:846]: Opps...my bad...this is a LAMDA.

**Review by rozumx — CHANGES_REQUESTED** [review]
Could the format changes requested and implemented for:
`src/runtime_src/core/tools/xbutil2/SubCmdExamine.cpp`
be also applied for:
`src/runtime_src/core/tools/xbmgmt2/SubCmdExamine.cpp`

**Review by rozumx — APPROVED** [review]
Thank you for making these changes.

**Pull request merged**